sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -125,10 +162,23 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
     val stagingDir = hadoopConf.get("hive.exec.stagingdir", ".hive-staging")
     val scratchDir = hadoopConf.get("hive.exec.scratchdir", "/tmp/hive")
 
+    // Hive sets session_path as HDFS_SESSION_PATH_KEY(_hive.hdfs.session.path) in hive config
+    val sessionScratchDir = externalCatalog.unwrapped.asInstanceOf[HiveExternalCatalog]
+      .client.getConf("_hive.hdfs.session.path", "")
+    logDebug(s"path '${path.toString}', staging dir '$stagingDir', " +
+      s"scratch dir '$scratchDir', session scratch dir '$sessionScratchDir' are used")
+
     if (hiveVersionsUsingOldExternalTempPath.contains(hiveVersion)) {
       oldVersionExternalTempPath(path, hadoopConf, scratchDir)
     } else if (hiveVersionsUsingNewExternalTempPath.contains(hiveVersion)) {
-      newVersionExternalTempPath(path, hadoopConf, stagingDir)
+      // HIVE-14270: Write temporary data to HDFS when doing inserts on tables located on S3
+      // Copied from Context.java#getTempDirForPath of Hive 2.3
+      if (isBlobStoragePath(hadoopConf, path)
+        && !useBlobStorageAsScratchDir(hadoopConf)) {
+        getMRTmpPath(hadoopConf, sessionScratchDir, scratchDir)
```

Comment:
Could you add a new conf under the name spark `spark.sql.hive.` for this new feature? Also, please follow the naming rule: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala#L20-L47

By doing so, [sql/create-docs.sh](https://github.com/apache/spark/blob/master/sql/create-docs.sh) could automatically generate a doc for the conf.

Reply:
btw, we need to use the same config name with hive?

Reply:
Although `I used spark.hadoop.hive.blobstore.*` in the initial commit, it was for consistency.
Since I do not have any technical reason here, I changed it to use SQLConf to follow the naming rule here.

Reply:
cc @dongjoon-hyun @wangyum